pkg/api/agentPoolOnlyApi/v20180331/validate.go

```diff
@@ -130,45 +133,117 @@ func validateUniqueProfileNames(profiles []*AgentPoolProfile) error {
+			return ErrorInvalidDNSServiceIP
+		}
+
+		// validate DockerBridgeCidr
+		_, _, err = net.ParseCIDR(n.DockerBridgeCidr)
+		if err != nil {
+			return ErrorInvalidDockerBridgeCidr
+		}
+
+		// validate DNSServiceIP is within ServiceCidr
+		if !serviceCidr.Contains(dnsServiceIP) {
+			return ErrorDNSServiceIPNotInServiceCidr
+		}
 
-			subIDMap[agentSubID] = subIDMap[agentSubID] + 1
-			resourceGroupMap[agentRG] = resourceGroupMap[agentRG] + 1
-			agentVNETMap[agentVNET] = agentVNETMap[agentVNET] + 1
+		// validate DNSServiceIP is not the first IP in ServiceCidr. The first IP is reserved for redirect svc.
+		if dnsServiceIP.String() == strings.TrimSuffix(serviceCidr.IP.String(), "0")+"1" {
+			return ErrorDNSServiceIPAlreadyUsed
 		}
 
-		// TODO: Add more validation to ensure all agent pools belong to the same VNET, subscription, and resource group
-		// 	if(len(subIDMap) != len(a.AgentPoolProfiles))
+		// validate custom VNET logic
+		if n.isCustomVNET() {
+			var subscription string
+			var resourceGroup string
+			var vnet string
+
+			if a.AgentPoolProfiles == nil {
+				return ErrorNilAgentPoolProfile
+			}
+
+			for _, agentPool := range a.AgentPoolProfiles {
+				if !agentPool.IsCustomVNET() {
+					return ErrorAgentPoolNoSubnet
+				}
 
-		// 	return errors.New("Multiple VNETS specified.  Each agent pool must reference the same VNET (but it is ok to reference different subnets on that VNET)")
-		// }
+				// TODO: validate VnetSubnetCidr
+
+				// TODO: validate dockerBridgeCidr is not within VnetSubnetCidr
+
+				// TODO: validate ServiceCidr does not overlap with VnetSubnetCidr
+
+				// TODO: validate VnetSubnetCidr does n ot overlap with other VnetSubnetCidr
+
+				// TODO: validate IP address pre-allocation does not exceed VnetSubnetCidr capacity
+
+				// validate subsricption, resource group and vnet are the same among subnets
+				subnetSubscription, subnetResourceGroup, subnetVnet, _, err := GetVNETSubnetIDComponents(agentPool.VnetSubnetID)
+				if err != nil {
+					return ErrorParsingSubnetID
+				}
+
+				if subscription == "" {
+					subscription = subnetSubscription
+				} else {
+					if subscription != subnetSubscription {
+						return ErrorSubscriptionNotMatch
+					}
+				}
+
+				if resourceGroup == "" {
+					resourceGroup = subnetResourceGroup
+				} else {
+					if resourceGroup != subnetResourceGroup {
+						return ErrorResourceGroupNotMatch
+					}
+				}
+
+				if vnet == "" {
+					vnet = subnetVnet
+				} else {
+					if vnet != subnetVnet {
+						return ErrorVnetNotMatch
+					}
+				}
+			}
+		}
+
+	} else if string(n.NetworkPlugin) == string(Kubenet) {
+		if n.ServiceCidr != "" || n.DNSServiceIP != "" || n.DockerBridgeCidr != "" {
+			return ErrorKubenetNoCustomization
+		}
+	} else {
+		return ErrorInvalidNetworkPlugin
 	}
 
 	return nil
 }
 
+func (n *NetworkProfile) isCustomVNET() bool {
+	// use ServiceCidr check to decide if Custom VNET is triggered
+	return n.ServiceCidr != ""
```

Comment:
If `n.ServiceCidr == ""` and `n.NetworkPlugin == string(azure)`, it won't be custom VNET? Tighten this validation so that both are true (`n.ServiceCidr != "" && n.NetworkPlugin == string(azure)`)

Reply:
As discussed, will use agentpool subnets to implement isCustomVNET function.